MozziGuts_impl_ARDUINOFSP.hpp

```diff
@@ -0,0 +1,227 @@
+#endif
+      }
+    }
+  }
+
+  dtc_info.transfer_settings_word_b.dest_addr_mode = TRANSFER_ADDR_MODE_FIXED;
+  dtc_info.transfer_settings_word_b.repeat_area = TRANSFER_REPEAT_AREA_SOURCE;
+  dtc_info.transfer_settings_word_b.irq = TRANSFER_IRQ_END;
+  dtc_info.transfer_settings_word_b.chain_mode = TRANSFER_CHAIN_MODE_DISABLED;
+  dtc_info.transfer_settings_word_b.src_addr_mode =
+    TRANSFER_ADDR_MODE_INCREMENTED;
+  dtc_info.transfer_settings_word_b.mode = TRANSFER_MODE_REPEAT;
+  dtc_info.p_src = (void const *)NULL;
+
+  dtc_info.num_blocks = 0;
+  dtc_info.length = 0;
+
+  dtc_cfg.p_info = &dtc_info;
+  dtc_cfg.p_extend = &dtc_cfg_extend;
+
+  dtc_cfg_extend.activation_source = FSP_INVALID_VECTOR;
+}
+
+void dac_init() {
+  if (IS_DAC_8BIT(pin)) {
+#if DAC8_HOWMANY > 0
+    if (GET_CHANNEL(pin) < DAC8_HOWMANY) {
+      _dac8[GET_CHANNEL(pin)].init();
+    }
+#endif
+  } else {
+    if (GET_CHANNEL(pin) < DAC12_HOWMANY) {
+      _dac12[GET_CHANNEL(pin)].init();
+    }
+  }
+}
+
+#endif
+
+
+//////////////// TIMER ////////////////
+
+#if EXTERNAL_AUDIO_OUTPUT == true
+void timer_callback_dummy(timer_callback_args_t __attribute__((unused)) *args){defaultAudioOutput();};
+#else
+//void timer_callback_dummy(timer_callback_args_t __attribute__((unused)) *args){
+void timer_callback_dummy(timer_callback_args_t __attribute__((unused)) *args){output_buffer.read();}; // to empty the buffer (the dac does not take care of it), a bit a waste of timer...
+#endif
+
+void timer_init() {
+  uint8_t type = 0;
+  int8_t tindex = FspTimer::get_available_timer(type);
+
+  if (tindex < 0) {
+    tindex = FspTimer::get_available_timer(type, true);
+  }
+
+  if (tindex >= 0) {
+    timer.begin(TIMER_MODE_PERIODIC, type, tindex, AUDIO_RATE, 50.0,timer_callback_dummy);
+    timer.setup_overflow_irq();
+  }
+  
+#if EXTERNAL_AUDIO_OUTPUT != true // we need to set up another timer for dac caring
+  // note: it is running at the same speed than the other one, but could not manage
+  // to get the other one updating the dac and removing the samples from the buffer…
+    tindex = FspTimer::get_available_timer(type);
+
+  if (tindex < 0) {
+    tindex = FspTimer::get_available_timer(type, true);
+  }
+
+  if (tindex >= 0) {
+    FspTimer::force_use_of_pwm_reserved_timer();
+    timer_dac.begin(TIMER_MODE_PERIODIC, type, tindex, AUDIO_RATE, 50.0);
+    timer_dac.setup_overflow_irq();    
+     dtc_cfg_extend.activation_source = timer_dac.get_cfg()->cycle_end_irq;
+     timer_dac.open();    
+#endif
+    timer.open();
+  }
+}
+
+
+inline void audioOutput(const AudioOutput f) {
+  output_buffer.write(f+AUDIO_BIAS);
+}
+#define canBufferAudioOutput() (!output_buffer.isFull())
+
+
+static void startAudio() {
+#if EXTERNAL_AUDIO_OUTPUT != true
+  dac_creation(AUDIO_CHANNEL_1_PIN);
+#endif
+  timer_init();
+#if EXTERNAL_AUDIO_OUTPUT != true
+  dac_init();
+  R_DTC_Open(&dtc_ctrl, &dtc_cfg);
+  R_DTC_Enable(&dtc_ctrl);
+
+  dtc_cfg.p_info->p_src = output_buffer.address();
```

Comment:
So, do I get this right? The DAC will keep looping over the buffer at AUDIO_RATE. Further down, below, you are advancing the read position as known to Mozzi, also at AUDIO_RATE.

Nifty, but please add a short explanation.

Reply:
Yeah, after searching for quite some time the best way to interact with this DAC I found out that seems to be the more satisfying (updating the buffers as a bunch is way to slow). Will add some explanation as this is deviating quite a lot from the previous strategies used in Mozzi.

Reply:
I left this one for when I'll have a lot of time in front of me (or someone else solves it, if possible). I remember trying quite hard to make this work to not avail and getting back into the not so commented source code of these timers always takes a bit of time…

Reply:
Added some documentation about the strategy used here as it is fairly different from before.